### PR TITLE
usm: Add support for Kafka Produce v9/v10

### PR DESCRIPTION
**Pull request by vitkyrka**
### What does this PR do?

Adds support for v9 and v10 of the Kafka Produce API to USM.

### Motivation

https://datadoghq.atlassian.net/browse/USMON-828

### CODE REVIEW

File: pkg/network/protocols/kafka/testdata/docker-compose.yml [hunk 1]
```diff
@@ -10,7 +10,7 @@ services:
     tmpfs:
       - /bitnami/zookeeper/data
   kafka:
-    image: bitnami/kafka:3.4
+    image: bitnami/kafka:3.7
```
Comment (guyarb) [inline pkg/network/protocols/kafka/testdata/docker-compose.yml:13]: you need to modify ami-builder repository

Comment (guyarb) [reply pkg/network/protocols/kafka/testdata/docker-compose.yml:13]: and system-probe load tests
https://github.com/DataDog/system-probe-test-environments/blob/4d92a3613129020dba3eef55226578e917f06e32/k8s/kafka-load-test/templates/kafka.yaml#L38C6-L38C43

File: pkg/network/usm/kafka_monitor_test.go [hunk 2]
```diff
@@ -119,8 +119,8 @@ func (s *KafkaProtocolParsingSuite) TestKafkaProtocolParsing() {
 	var versions []*kversion.Versions
 	versions = append(versions, kversion.V2_5_0())
 
-	fetch12 := kversion.V3_4_0()
-	fetch12.SetMaxKeyVersion(kafka.ProduceAPIKey, 8)
+	fetch12 := kversion.V3_7_0()
```
Comment (DanielLavie) [inline pkg/network/usm/kafka_monitor_test.go:122]: Consider rename `fetch12` as it also includes the ProduceAPIKey

**Review by DanielLavie — APPROVED** [review]

File: pkg/network/ebpf/c/protocols/kafka/kafka-classification.h [hunk 3]
```diff
@@ -165,9 +164,37 @@ static __always_inline bool skip_varint_number_of_topics(pktbuf_t pkt, u32 *offs
     return true;
 }
 
+static __always_inline __maybe_unused bool skip_varint(pktbuf_t pkt, u32 *offset, u32 max_bytes) {
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/kafka/kafka-classification.h:167]: nit: worth adding documentation to the function

File: pkg/network/ebpf/c/protocols/kafka/kafka-classification.h [hunk 4]
```diff
@@ -165,9 +164,37 @@ static __always_inline bool skip_varint_number_of_topics(pktbuf_t pkt, u32 *offs
...
+    u8 current_byte = bytes[0];
+
+    #pragma unroll
+    for (; i < max_bytes; i++) {
+        *offset += 1;
+        if (!isMSBSet(current_byte)) {
+            break;
+        }
+        current_byte = bytes[i];
+    }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/kafka/kafka-classification.h:186]: Does pragma unroll work good with "a variable" length?
is it a "replacement" for macros?

File: pkg/network/ebpf/c/protocols/kafka/kafka-classification.h [hunk 5]
```diff
@@ -165,9 +164,37 @@ static __always_inline bool skip_varint_number_of_topics(pktbuf_t pkt, u32 *offs
...
 }
 
+static __always_inline __maybe_unused bool skip_varint(pktbuf_t pkt, u32 *offset, u32 max_bytes) {
+    u8 bytes[4] = {};
+
+    if (*offset + max_bytes > pktbuf_data_end(pkt)) {
+        return false;
+    }
+
+    pktbuf_load_bytes(pkt, *offset, bytes, max_bytes);
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/kafka/kafka-classification.h:174]: later in the code, you assume `max_bytes` is smaller than 4, as you're accessing indices from 1 to max bytes

File: pkg/network/ebpf/c/protocols/kafka/kafka-classification.h [hunk 6]
```diff
@@ -165,9 +164,37 @@ static __always_inline bool skip_varint_number_of_topics(pktbuf_t pkt, u32 *offs
...
+
+    u32 i = 1;
+    u8 current_byte = bytes[0];
+
+    #pragma unroll
+    for (; i < max_bytes; i++) {
+        *offset += 1;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/kafka/kafka-classification.h:181]: nit, having this line in the for-loop is redundant
you can simple do `*offset += i` after the for-loop

File: pkg/network/ebpf/c/protocols/kafka/kafka-classification.h [hunk 7]
```diff
@@ -165,9 +164,37 @@ static __always_inline bool skip_varint_number_of_topics(pktbuf_t pkt, u32 *offs
...
+            break;
+        }
+        current_byte = bytes[i];
+    }
+
+    if (i == max_bytes && isMSBSet(current_byte)) {
+        return false;
+    }
+
+    return true;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/kafka/kafka-classification.h:192]: nit, can be simplified to `return !(i == max_bytes && isMSBSet(...)`

File: pkg/network/ebpf/c/protocols/kafka/kafka-classification.h [hunk 8]
```diff
@@ -228,12 +255,40 @@ static __always_inline bool validate_first_topic_name(pktbuf_t pkt, bool flexibl
     CHECK_STRING_VALID_TOPIC_NAME(TOPIC_NAME_MAX_STRING_SIZE_TO_VALIDATE, topic_name_size, topic_name);
 }
 
+static __always_inline bool skip_request_tagged_fields(pktbuf_t pkt, u32 *offset) {
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/kafka/kafka-classification.h:258]: nit: missing documentation

File: pkg/network/ebpf/c/protocols/kafka/kafka-classification.h [hunk 9]
```diff
@@ -228,12 +255,40 @@ static __always_inline bool validate_first_topic_name(pktbuf_t pkt, bool flexibl
...
+    if (*offset >= pktbuf_data_end(pkt)) {
+        return false;
+    }
+
+    u8 num_tagged_fields = 0;
+
+    pktbuf_load_bytes(pkt, *offset, &num_tagged_fields, 1);
+    *offset += 1;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/kafka/kafka-classification.h:266]: nit, use `sizeof(num_tagged_fields)`

File: pkg/network/ebpf/c/protocols/kafka/kafka-classification.h [hunk 10]
```diff
@@ -228,12 +255,40 @@ static __always_inline bool validate_first_topic_name(pktbuf_t pkt, bool flexibl
...
+}
+
 // Getting the offset (out parameter) of the first topic name in the produce request.
 static __always_inline bool get_topic_offset_from_produce_request(const kafka_header_t *kafka_header, pktbuf_t pkt, u32 *out_offset) {
     const s16 api_version = kafka_header->api_version;
     u32 offset = *out_offset;
+    bool flexible = false;
+
+    if (api_version >= 9) {
+        flexible = true;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/kafka/kafka-classification.h:279]: nit: `flexible = api_version >= 9`

File: pkg/network/ebpf/c/protocols/kafka/kafka-classification.h (hunk 3 above)
Comment (vitkyrka) [reply pkg/network/ebpf/c/protocols/kafka/kafka-classification.h:175]: Done.

**Pull request merged**